### ISSUE TITLE
Play around with support for multiple windows

### ISSUE THREAD

**Issue by jlnr** [gosu/gosu#387]
Right now, it is technically possible to `show` two windows at the same time by using the `Window#tick` method. Example:

```ruby
w1, w2 = Window.new(300, 200), Window.new(300, 200)
puts "wat" while w1.tick or w2.tick
```

However, this does not really work because all `Gosu::Window` instances share the same `SDL_Window`. This wasn't really a big deal before `Window#tick`, because the only way to present a window was a blocking call to `#show`.

Considering that #287 was also complicated by `Window#tick`, maybe it is time to evaluate whether `#tick` is really worth all the trouble (e.g. see if Gosu can actually be used with a GUI toolkit now?); and maybe it is worth it just for this particular use case, namely, showing two windows at the same time?

Bonus challenge: The above must run at 60 FPS, the two `#tick` (`#draw`) must not *both* block for vsync.

**Comment by pmer** [comment]
Just wondering... having a hard time seeing the interaction of #287 and `Window#tick`... What happened there?

**Comment by pmer** [comment]
Also, the vsync issue could be achieved by going the very un-Gosu-like route 😮 of having a global `Gosu::tick` function that ticks each Gosu window and only syncs on the last.

**Comment by jlnr** [comment]
The problem with #287 was this: When `close` is called from within `tick` (in `update` or any of the other callbacks), `tick` must return false and the window must disappear; when `close` is called outside of `tick`, then the next call to `tick` should return `false`. Or at least that is the only behaviour that I thought was consistent with the spirit of `tick` (`show() === while(tick())`.

**Comment by pmer** [comment]
Oh, that's very interesting. Thank you.

I'm looking at the 0.10.2 release and it says that `tick` is experimental... would it be reasonable to question its type signature? Because `tick` seems to have multiple responsibilities that could be split up if we can do that.

It:

- Opens windows
- Runs ticks
- Returns some value derived from the current state

Theoretically, these could be one function each, for a total of three functions. Although I'm not going to advocate for or against that right now because I haven't thought it through. 🙂

**Comment by pmer** [comment]
I think the third window state of `CLOSING` would not be necessary if the `tick` function only had two of the three (or one of the three 🙂) responsibilities.

**Comment by Kjarrigan** [comment]
> Considering that #287 was a bit hairy because of the existence of Window#tick, maybe it is time to evaluate whether #tick is really worth all the trouble? Can Gosu actually be used (reliably) with a GUI toolkit now?

What about starting a poll or thread or whatever in /r/gosu and the forums to ask the loyal followers about their usecases for gosu, platforms used,  most common gem combinations and maybe some of the open "enhancement" issues and most requested wishes (I vote for screenshots :smile:). I think we could get some valuable feedback. Maybe even consider adding a little gem-postinstall-message with a link to the open discussion/poll for a few weeks.

**Comment by pmer** [comment]
If that is possible, any leads we find into this issue from that would be better than us trying to imagine what a user *might* want.

**Comment by jlnr** [comment]
> tick seems to have multiple responsibilities

Agreed, it's definitely a confusing API. On the other hand, I like that it hides an advanced use case behind a single method that most people will ignore.

There have been several requests for a way to integrate Gosu with existing GUI toolkits, so for `tick` the question is not whether anyone ever requested it, but whether there is any GUI toolkit for Ruby that works *and* works with Gosu now. If not, it should be removed and [Spooner's fidgit](https://github.com/gosu/fidgit) should be cleaned up a little so that it can be used for quick level editors etc.

**Comment by jlnr** [comment]
I've changed the issue title to be more general because using two Window instances in parallel is awkward right now even without `tick` - for example, `#caption=` modifies the same shared state on all instances.

Maybe the easy way forward is to print a warning when creating more than one window. Now that windows can be resized, I can't think of a use case for that anymore.

**Comment by jlnr** [comment]
Re: Poll, good idea. Do you feel that the current backlog contains too many new features? Should we wait for someone to explicitly request #378, for example? Or should it mostly be a retrospective poll about which old features can be safely dropped? For example, I really want to kill support for Ruby 1.8 and 1.9, but I remember that lol_o2/KoBeWi is stuck on Ruby 1.9.3 because of Ashton. Ouch.

@Kjarrigan: You don't even need to vote for screenshots because everything that helps the testing infrastructure *and* absorbs functionality from Ashton sounds good 👍 (I've updated #317 which is pretty much that.)

**Comment by Kjarrigan** [comment]
> Do you feel that the current backlog contains too many new features?

Thats not that easy to answer. At the moment there are 36 open issues labeled with enhancement  some of them a couple months or even years old, so maybe we should try to focus on solving these first, but on the other hand newer request might be more "needed" than something no one asked again for months. I really want to solve some of the issues but my lack of C++ knowledge restricts me quite a bit, but I hope to get some stuff done nevertheless :smile:

> I really want to kill support for Ruby 1.8 and 1.9, but I remember that lol_o2/KoBeWi is stuck on Ruby 1.9.3 because of Ashton.

Dropping support for Ruby < 2.0 sounds good to me. I have never used Ashton before - is there significant problem on making it Ruby 2.x compatible? Thats could be a focus as well. 

> Should we wait for someone to explicitly request #378 again, for example?

Yeah. This specific problem I might wan't to finish, but the poll should be more like: Which 3 features do you want most for Gosu? (Can be specific like Screenshots or more general like add more Audio capabilites)
Or which gems do you use most with Gosu (e.g. chingu, asthon, texplay, ...)
If everyone uses asthon for example we might consider improving this first or port some of its features to the "Gosu core".

Just thoughts - your the one in charge ;-) Maybe we could even directly ask some of the "active" members who actually created games/gems with/around Gosu like shawn42, lol_o2, KoBeWi, djzk, Spooner, 	Dahrkael.

**Comment by jlnr** [comment]
It would probably be good to also reach out to people who use Gosu for teaching. What are their pain points? These people don't seem to ever visit the forums or reddit, so I have no idea. I just see [posts like this](http://stackoverflow.com/questions/19886147/how-do-i-install-gosu) that pop up from time to time :)
